GAE/src/org/waterforpeople/mapping/app/web/DataProcessorRestServlet.java

```diff
@@ -1443,13 +1443,27 @@ private void deleteCascadeNodes(Long cascadeResourceId, Long parentNodeId) {
         List<CascadeNode> nodes = dao.listCascadeNodesByResourceAndParentId(cascadeResourceId,
                 parentNodeId == null ? 0l : parentNodeId);
 
-        for (CascadeNode node : nodes) {
-            scheduleCascadeNodeDeletion(cascadeResourceId, node.getKey().getId());
+        if (!areLeafNodes(dao, cascadeResourceId, nodes)) {
+            for (CascadeNode node : nodes) {
+                scheduleCascadeNodeDeletion(cascadeResourceId, node.getKey().getId());
+            }
         }
 
         dao.delete(nodes);
     }
 
+    private boolean areLeafNodes(CascadeNodeDao dao, Long cascadeResourceId,
+            List<CascadeNode> nodes) {
+        if (nodes.size() == 0) {
+            return true;
```

Comment:
Perhaps you can move the `nodes.size() == 0` branch to just before the `if(!areLeafNodes())`.  This way the meaning of the `areLeafNodes()` function is not mixed with checking if the list of nodes is empty.